cfy_manager/components/globals.py

```diff
@@ -93,69 +95,80 @@ def _apply_forced_settings():
         config[POSTGRESQL_CLIENT][SSL_ENABLED] = True
 
 
-class Listener(TypedDict, total=False):
+@dataclass(slots=True)
+class Listener:
     host: str
-    port: int
-    server_name: str
-    ssl: bool
-    cert_path: str
-    key_path: str
-
-
-def _format_listener(listener) -> Listener:
-    if not listener:
-        listener = {'host': 'localhost'}
-    elif isinstance(listener, str):
-        listener = {'host': listener}
-
-    formatted = listener.copy()
-    if not formatted.get('port'):
-        formatted['port'] = config[MANAGER]['external_rest_port']
-    if not formatted.get('ssl'):
-        formatted['ssl'] = config[MANAGER][SECURITY]['ssl_enabled']
-    return formatted
-
+    port: Optional[int] = None
+    server_name: Optional[str] = '_'
+    ssl: Optional[bool] = False
+    cert_path: Optional[str] = None
+    key_path: Optional[str] = None
+
+    @staticmethod
+    def _validate_config_listener(listener: dict | str):
+        if not listener:
+            raise BootstrapError(f'listeners: got null value: {listener!r}')
+
+    @classmethod
+    def from_config(cls, original: dict | str, config: dict) -> 'Listener':
+        if isinstance(original, str):
+            config_listener = {'host': original}
+        else:
+            config_listener = original
+
+        listener = cls(**original)
+        if listener.port is None:
+            listener.port = config[MANAGER]['external_rest_port']
+        if listener.ssl is None:
+            listener.ssl = config[MANAGER][SECURITY]['ssl_enabled']
+        return listener
+
+    def __hash__(self):
```

Comment:
You can skip defining __hash__ and __eq__ if you set unsafe_hash = True (unless that's not needed with slots=True?) and then define the non-hash/eq parts using this approach: https://github.com/cloudify-cosmo/cloudify-manager/blob/master/rest-service/manager_rest/rest/responses.py#L164